druid-shell/src/text_input.rs

```diff
@@ -0,0 +1,516 @@
+    ///
+    /// This is automatically implemented, but you can override this if you have some faster system to determine string length.
+    fn utf16_to_utf8(&mut self, utf16_range: Range<usize>) -> usize {
+        if utf16_range.is_empty() {
+            return 0;
+        }
+        let doc_range = 0..self.len();
+        let text = self.slice(doc_range);
+        let utf16: Vec<u16> = text
+            .encode_utf16()
+            .skip(utf16_range.start)
+            .take(utf16_range.end)
+            .collect();
+        String::from_utf16_lossy(&utf16).len()
+    }
+
+    /// Replaces a range of the text document with `text`.
+    /// If `range.start` or `range.end` do not fall on a code point sequence boundary, this method may panic.
+    /// Equivalent to `String::replace_range`.
+    ///
+    /// This method also sets the composition range to `None`, and updates the selection:
+    ///
+    /// - If both the selection's anchor and extent are `< range.start`, then nothing is updated.
+    /// - If both the selection's anchor and extent are `> range.end`, then subtract `range.len()` from both, and add `text.len()`.
+    /// - If neither of the previous two conditions are true, then set both anchor and extent to `range.start + text.len()`.
+    ///
+    /// After the above update, if we increase each end of the selection if necessary to put it on a grapheme cluster boundary.
+    ///
+    /// Requries a mutable lock.
+    fn replace_range(&mut self, range: Range<usize>, text: &str);
+
+    /// Given a `Point`, determine the corresponding text position.
+    fn hit_test_point(&mut self, point: Point) -> HitTestPoint;
+
+    /// Returns the character range of the line (soft- or hard-wrapped) containing the character
+    /// specified by `char_index`.
+    fn line_range(&mut self, char_index: usize, affinity: Affinity) -> Range<usize>;
+
+    /// Returns the bounding box, in window coordinates, of the visible text document. For instance,
+    /// a text box's bounding box would be the rectangle of the border surrounding it, even if the text box is empty.
+    /// If the text document is completely offscreen, return `None`.
+    fn bounding_box(&mut self) -> Option<Rect>;
+
+    /// Returns the bounding box, in window coordinates, of the range of text specified by `range`.
+    /// Ranges will always be equal to or a subrange of some line range returned by `TextInputHandler::line_range`.
+    /// If a range spans multiple lines, `slice_bounding_box` may panic.
+    fn slice_bounding_box(&mut self, range: Range<usize>) -> Option<Rect>;
+
+    /// Applies some special action to the text field.
+    ///
+    /// Requries a mutable lock.
+    fn handle_action(&mut self, action: TextInputAction);
+}
+
+#[allow(dead_code)]
+/// Simulates `TextInputHandler` calls on `handler` for a given keypress `event`. This circumvents the platform, and so can't
```

Comment:
Rustdoc treats doc comments kind of like git treats commit messages; the first line is 'special', and acts as a sort of summary; for instance if you're in the docs for a module, where it lists the contents of that module, the first line is displayed beside each item.

In practice what this means is it makes sense to try and have a simple first line (like the first sentence, here) and then a newline followed by the more detailed docs.

Reply:
I think this should be fixed, but if you spot a long first line I've missed somewhere, let me know!